helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java

```diff
@@ -0,0 +1,256 @@
+import org.apache.helix.rest.metadatastore.constant.MetadataStoreRoutingConstants;
+import org.apache.helix.rest.metadatastore.exceptions.ZkLockException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class ZkRoutingDataWriter implements MetadataStoreRoutingDataWriter {
+  private static final Logger LOG = LoggerFactory.getLogger(ZkBaseDataAccessor.class);
+
+  private final String _namespace;
+  private final String _zkAddress;
+  private final HelixZkClient _zkClient;
+  private final ZkDistributedLeaderElection _leaderElection;
+
+  public ZkRoutingDataWriter(String namespace, String zkAddress) {
+    if (namespace == null || namespace.isEmpty()) {
+      throw new IllegalArgumentException("namespace cannot be null or empty!");
+    }
+    _namespace = namespace;
+    if (zkAddress == null || zkAddress.isEmpty()) {
+      throw new IllegalArgumentException("Zk address cannot be null or empty!");
+    }
+    _zkAddress = zkAddress;
+    _zkClient = DedicatedZkClientFactory.getInstance()
+        .buildZkClient(new HelixZkClient.ZkConnectionConfig(zkAddress),
+            new HelixZkClient.ZkClientConfig().setZkSerializer(new ZNRecordSerializer()));
+
+    // Ensure that ROUTING_DATA_PATH exists in ZK. If not, create
+    // create() semantic will fail if it already exists
+    try {
+      _zkClient.createPersistent(MetadataStoreRoutingConstants.ROUTING_DATA_PATH);
+    } catch (ZkNodeExistsException e) {
+      // This is okay
+    }
+
+    // Get the hostname (REST endpoint) from System property
+    ZNRecord myServerInfo = new ZNRecord("dummy hostname");
+    _leaderElection = new ZkDistributedLeaderElection(_zkClient,
+        MetadataStoreRoutingConstants.LEADER_ELECTION_ZNODE, myServerInfo);
+  }
+
+  @Override
+  public synchronized boolean addMetadataStoreRealm(String realm) {
+    if (_leaderElection.isLeader()) {
+      if (_zkClient.isClosed()) {
+        throw new IllegalStateException("ZkClient is closed!");
+      }
+      return createZkRealm(realm);
+    }
+
+    // TODO: Forward the request to leader
+    return true;
+  }
+
+  @Override
+  public synchronized boolean deleteMetadataStoreRealm(String realm) {
+    if (_leaderElection.isLeader()) {
+      if (_zkClient.isClosed()) {
+        throw new IllegalStateException("ZkClient is closed!");
+      }
+      return _zkClient.delete(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + realm);
+    }
+
+    // TODO: Forward the request to leader
+    return true;
+  }
+
+  @Override
+  public synchronized boolean addShardingKey(String realm, String shardingKey) {
+    if (_leaderElection.isLeader()) {
+      if (_zkClient.isClosed()) {
+        throw new IllegalStateException("ZkClient is closed!");
+      }
+      // If the realm does not exist already, then create the realm
+      if (!_zkClient.exists(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + realm)) {
```

Comment:
Can you create a variable for the path so you don't need to repeatedly type it out?

Reply:
Will do.